### PR TITLE
13374-create-documentation-describing-folder-structure-and-usage-best-practices

### PR DESCRIPTION
**Pull request by frolui**

### CODE REVIEW

File: DOCUMENTATION.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+- [procedures/](procedures) - service SQL procedures, used in more than one other file
+- [scripts/](scripts) - scripts that perform data transformation on top of table without creating new table
+- [tables/](tables) - SQL, which generates a table with the same name as the script
+- [static_data](static_data) - static file-based data stored in geocint repository
+All these folders are removed and recreated each time when geocint pipeline starts.
+
+After running the pipeline, make will create additional folders. These folders are used to store input (in folder), intermediate (mid folder), and output (out folder) data files.
+Make-profiler uses the timestamps of these files as timestamps for the execution of such a target:
+- [data/](data) - file-based input, middle, output data.
+	- data/in - all input data, downloaded elsewhere
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:36]: but planet.psm.pbf not in data/in it is in data

File: DOCUMENTATION.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+2. Clone 3 repository (geocint-runner, geocint-openstreetmap, your repo) to ~/
+3. The geocint pipeline should send messages to the Slack channel. Create a channel, generate Slack token and store it in the `SLACK_KEY` variable in file `~/.profile`.
+```shell
+export SLACK_KEY=<your_key>
+```
+4. Copy [config.inc.sh.sample](config.inc.sh.sample) from geocint-runner to ~/:
+```shell
+cp ~/geocint-runner/config.inc.sh.sample ~/config.inc.sh
+```
+Open ~/config.inc.sh and set the necessary values of variables.
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:68]: maybe move about SLACK_KEY in this section?

File: DOCUMENTATION.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+5. Run installers:
+- ~/geocint-runner/runner_install.sh (necessary dependencies to run runner part)
+- ~/geocint-openstreetmap/openstreetmap_install.sh (necessary dependencies to run runner part)
+- ~/geocint-runner/set_mods.sh
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:72]: what does this script do?

File: DOCUMENTATION.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+deploy/s3/fire_stations: data/out/fire_stations/fire_stations_h3_r8_count.gpkg.gz | deploy/s3 ## deploy/s3 is a dependency from geocint-runner Makefile
+    aws s3 cp data/out/fire_stations/fire_stations_h3_r8_count.gpkg.gz \
+   	 s3://your_s3_bucket/fire_stations_h3_r8_count.gpkg.gz \
+   	 --profile your_s3_profile \
+   	 --acl public-read
+    touch $@
+```
+As a result of this target execution, we don't have a file, which means we have to use the touch $@ line at the end. This command will create an empty deploy/s3/fire_stations file whose creation timestamp will be used by Make-profiler.
+
+if you want to rebuild some targets chain each time please see How geocint pipeline workyou run the pipeline, you must add the initial target of this chain to the TARGETS_TO_CLEAN variable in the configuration file config.inc.sh).
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:132]: grammar errors

File: DOCUMENTATION.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+- [logs/](logs) - files - files with targets execution logs
+- [report/](report) - folder to store HTML reports.
+
+These folders are not deleted or re-created each time the geocint pipeline is run, to avoid rebuilding targets that should not be rebuilt every time (if you want to rebuild some targets chain each time please see How geocint pipeline).
+You shouldn’t store all your input datasets in data/in/ folder. To make your data storage more organized, you can create additional folders for separate data sources (for example data/in/source_name). This also applies to other catalogs.
+
+## Geocint open-source installation and first run guide
+
+### Intallation
+Before to install your own geocint pipeline instance, you should create a repository to store your own part of the pipeline.
```
[the remark this answers is not in the document]
Comment (frolui) [reply DOCUMENTATION.md:52]: fixed

File: DOCUMENTATION.md (hunk 1 above)
Comment (frolui) [reply DOCUMENTATION.md:36]: this is how it was originally designed

File: DOCUMENTATION.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+- [geocint-private] any repository that contains your additional functionality
+
+During the deployment process, you should clone all these repos to ~/.
+Then start_geocint.sh will copy files from all these repositories to ~/geocint folder. Geocint folder will be the working folder for geocint pipeline
+
+In general case ~/geocint folder includes the next folders:
+- [functions/](functions) - service SQL functions, used in more than one other file
+- [procedures/](procedures) - service SQL procedures, used in more than one other file
+- [scripts/](scripts) - scripts that perform data transformation on top of table without creating new table
+- [tables/](tables) - SQL, which generates a table with the same name as the script
```
[the remark this answers is not in the document]
Comment (frolui) [reply DOCUMENTATION.md:29]: removed

File: DOCUMENTATION.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+to database OpenStreetMap planet dump
+- [geocint-private] any repository that contains your additional functionality
+
+During the deployment process, you should clone all these repos to ~/.
+Then start_geocint.sh will copy files from all these repositories to ~/geocint folder. Geocint folder will be the working folder for geocint pipeline
+
+In general case ~/geocint folder includes the next folders:
+- [functions/](functions) - service SQL functions, used in more than one other file
+- [procedures/](procedures) - service SQL procedures, used in more than one other file
+- [scripts/](scripts) - scripts that perform data transformation on top of table without creating new table
```
[the remark this answers is not in the document]
Comment (frolui) [reply DOCUMENTATION.md:28]: removed

File: DOCUMENTATION.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+If you do not want to rebuild this chain, just make sure that the final target from your chain doesn’t depend on targets from the TARGETS_TO_CLEAN variable. This method can be useful when you have a large dataset that has not been updated recently and requires heavy and time-consuming pre-processing. In such a case, you probably want to prepare the data and load it into the database once and not repeat this work without changing the input data.
+But! Keep in mind, that dependencies inherited recursively, even if you don’t put data/in/some.tiff in the data/out/output.geojson.gz dependencies list, them will be data/out/output.geojson.gz’s dependency. This means that these two lines are equivalent:
+```
+data/out/output.geojson.gz: db/table/table_from_csv
+data/out/output.geojson.gz: data/in/some.tiff db/table/table_from_csv db/table/table_from_csv | data/out
+```
+You can read more about target's dependencies(prerequisites) in the official [GNU Make manual](https://www.gnu.org/software/make/manual/make.html#Prerequisite-Types)
+
+### User schemas
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:149]: in database

File: DOCUMENTATION.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+User schemas can be used for separate pipeline and dev data.
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:151]: I do not think that mapaction need this part at all.

File: DOCUMENTATION.md (hunk 3 above)
Comment (frolui) [reply DOCUMENTATION.md:72]: added explanation

File: DOCUMENTATION.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+`local   gis +geocint_users  trust`
+
+### How to analyse build time for tables
+
+Logs for every build are stored in `/home/gis/geocint/logs`
+
+This command can show lastN {*Total times in ms*} for some {*tablename*} ordered by date
+
+```bash
+cd /home/gis/geocint/logs
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:168]: this can be deleted and added as one command in find

File: DOCUMENTATION.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+find . -type f -regex ".*/db/table/osm_admin_boundaries/log.txt" -mtime -50 -printf "%T+ %p; " -exec awk '/Time:/ {sum += $4} END {print sum/60000 " min"}' '{}' \; | sort
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:169]: find /home/gis/geocint/logs -type f -regex ......

File: DOCUMENTATION.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+### How start_geocint.sh works
+
+After start_geocint.sh is run, it exports the variables from the configuration file ~/config.inc.sh. 
+It will then check the update flags in ~/config.inc.sh and git pull the repositories that have the flag set to “true”. 
+Next, it will merge geocint-runner, geocint-openstreetmap and your personal repository into one folder - you can set the name of this folder in ~/config.inc.sh file with $GENERAL_FOLDER variable. 
+Next, start_geocint.sh generates build and clean targets and adds include instructions. 
+After these events are completed, start_geocint.sh will launch the targets specified in the $RUN_TARGETS variable. The last step is to create/update the make.svg file containing the dependency graph.
+
+### How to write targets
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:116]: add requirement about each target need to have comment!

File: DOCUMENTATION.md [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+`-mtime -50` - collects every row from 50 days ago to now
+
+`-regex ".*/db/table/osm_admin_boundaries/log.txt"` - change `osm_admin_boundaries` to your {*tablename*}
+
+
+## Best practices of using
+
+There are a few simple rules, follow them to avoid troubles during the creation of your own pipeline:
+* Don’t create targets with a name which already exists in Makefile from geocint-runner repository, osm-make Makefile from geocint-openstreetmap repository or in your own Makefile;
+* Always add a short comment to your target (explain what hi does) - it’s a requirement;
```
Comment (lgudyma-kontur) [inline DOCUMENTATION.md:181]: (explain what **it** does)

Comment (frolui) [reply DOCUMENTATION.md:181]: fixed

File: DOCUMENTATION.md (hunk 11 above)
Comment (frolui) [reply DOCUMENTATION.md:169]: replaced

File: DOCUMENTATION.md (hunk 10 above)
Comment (frolui) [reply DOCUMENTATION.md:168]: replaced

File: DOCUMENTATION.md (hunk 9 above)
Comment (frolui) [reply DOCUMENTATION.md:151]: we should discuss it

File: README.md [hunk 14]
```diff
@@ -39,162 +39,4 @@ https://tldp.org/LDP/abs/html/
 - python3 is used for small tasks like unpivoting source data.
```
Comment (lgudyma-kontur) [inline README.md:39]: @kalenikaliaksandr proposed to use everywhere **python** not python3

File: DOCUMENTATION.md (hunk 8 above)
Comment (frolui) [reply DOCUMENTATION.md:149]: replaced

File: README.md (hunk 14 above)
Comment (frolui) [reply README.md:39]: replaced

File: config.inc.sh.sample [hunk 15]
```diff
@@ -21,9 +25,10 @@ UPDATE_OSM_LOGIC=false
 UPDATE_PRIVATE=true
 
 # targets block
-# list of dependencies for target "all"
-ALL_TARGETS='db/table/osm'
-# list of targets that will be runned
+# list of dependencies for target "all" - [FINAL] Meta-target on top of all other targets, or targets on parking
+# add here the final target of the chains that you want to execute not regularly, but only in certain cases
+ALL_TARGETS=''
```
Comment (lgudyma-kontur) [inline config.inc.sh.sample:30]: can this var be empty?

File: runner-install.sh [hunk 16]
```diff
@@ -8,4 +8,5 @@ sudo pip3 install pandas
 sudo apt-get install -y cmake
 sudo apt install -y unzip
 sudo apt-get -y install python3-boto3 python3-botocore # amazon.aws.aws_s3
```
Comment (lgudyma-kontur) [inline runner-install.sh:10]: we need to install python3 as python

File: scripts/set_modes.sh [hunk 17]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (lgudyma-kontur) [inline scripts/set_modes.sh:1]: what does this script do?

File: start_geocint.sh [hunk 18]
```diff
@@ -60,34 +60,18 @@ fi
...
+cd ~/
+# Merge geocint-runner, geocint-openstreetmap and your private repo to one folder and check duplicated files
+# This script use IGNORE_EXISTED_FILE variable from confic.inc.sh (by default ignore README.md and LICENSE files in root of every repo)
+copy_message="$(python3 geocint-runner/scripts/merge_repos_and_check_duplicates.py geocint-runner geocint-openstreetmap $PRIVATE_REPO_NAME)"
+
+# Chec if the message starts with "Copy.." in another case send message and exit
+if [ "$(echo $copy_message | head -c 1)" = "C" ]
 then
-  echo "Copy from geocint-runner, geocint-openstreetmap and $PRIVATE_REPO_NAME to geocint folder completed successfully" | python3 ~/geocint-runner/scripts/slack_message.py $SLACK_CHANNEL "$SLACK_BOT_NAME" $SLACK_BOT_EMOJI
+   echo $copy_message | python3 ~/geocint-runner/scripts/slack_message.py $SLACK_CHANNEL "$SLACK_BOT_NAME" $SLACK_BOT_EMOJI
```
Comment (lgudyma-kontur) [inline start_geocint.sh:71]: we need to use python everywhere

**Review by lgudyma-kontur — COMMENTED** [review]
left some comments

File: config.inc.sh.sample [hunk 19]
```diff
@@ -33,17 +38,25 @@ SLACK_CHANNEL='Your_slack_channel'
...
-# additional packages that should be installed
-TO_INSTALL='sudo pip3 install geopandas'
+# slack integration token, replace 'your_slack_key' with your slack integration key
+SLACK_KEY='your_slack_key'
 
 # clear target - Cleans the worktree for next nightly run. Does not clean non-repeating targets.
+# store directories that should be removed before each run, use whitespace ' ' as a delimiter to add directory
 RM_DIRECTORIES='data/planet-is-broken'
+# store targets that should be cleaned before each run, use whitespace ' ' as a delimiter to add target
 TARGETS_TO_CLEAN='data/planet-latest-updated.osm.pbf'
```
[the remark this answers is not in the document]
Comment (frolui) [reply config.inc.sh.sample:48]: can, but if we will remove data/planet-latest-updated.osm.pbf osm update process will not work

File: runner-install.sh (hunk 16 above)
Comment (frolui) [reply runner-install.sh:10]: added sudo apt install -y python-is-python3 as a convenient way to make symlink python to python3

File: DOCUMENTATION.md (hunk 2 above)
Comment (frolui) [reply DOCUMENTATION.md:68]: added comment "# this script is using one during installation to create /public_html and /domlogs folders and set access modes"

File: start_geocint.sh (hunk 18 above)
Comment (frolui) [reply start_geocint.sh:71]: replaced

File: scripts/set_modes.sh (hunk 17 above)
Comment (frolui) [reply scripts/set_modes.sh:1]: added comment "# this script is using one during installation to create /public_html and /domlogs folders and set access modes"

File: config.inc.sh.sample [hunk 20]
```diff
@@ -21,9 +25,10 @@ UPDATE_OSM_LOGIC=false
...
 
 # targets block
-# list of dependencies for target "all"
-ALL_TARGETS='db/table/osm'
-# list of targets that will be runned
+# list of dependencies for target "all" - [FINAL] Meta-target on top of all other targets, or targets on parking
+# add here the final target of the chains that you want to execute not regularly, but only in certain cases
+ALL_TARGETS=''
+# list of targets that will be runned automatically
 RUN_TARGETS='your_final_target'
```
[the remark this answers is not in the document]
Comment (frolui) [reply config.inc.sh.sample:32]: we can't specify which target you are going to run, so just leave a clear placeholder here

File: config.inc.sh.sample [hunk 21]
```diff
@@ -1,10 +1,14 @@
...
 # variable to store additional paths, that should be added to PATH variable
+# replase "gis" with your user name
 PATH_ARRAY='/home/gis/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:'
 # directory for pipeline execution 
 GENERAL_FOLDER=geocint
 # directories that should be added to env
-OPTIONAL_DIRECTORIES='.'
+OPTIONAL_DIRECTORIES=''
 # name of your private repositry
 PRIVATE_REPO_NAME=geocint-private
```
[the remark this answers is not in the document]
Comment (frolui) [reply config.inc.sh.sample:13]: i'm not sure, that we should replace it

File: DOCUMENTATION.md (hunk 2 above)
Comment (frolui) [reply DOCUMENTATION.md:68]: @lgudyma-kontur what do you mean

Comment (lgudyma-kontur) [reply DOCUMENTATION.md:68]: there is **3.** about setting slack_key, and later you there is setting vars in config.inc.sh... maybe join them in one

File: your_make.sample [hunk 22]
```diff
@@ -1,4 +1,32 @@
...
+
+# these makefiles stored in geocint-runner and geocint-openstreetmap repositories
+# runner_make contains basic set of targets for creation project folder structure
+# osm_make contains contain set of targets for osm data processing
+include runner_make osm_make
+
+## ------------- CONTROL BLOCK -------------------------
+
+# replace your_final_target placeholder with the names of final target, that you will use to run pipeline
+# you can also add here the names of targets that should not be rebuilt automatically, but only when conditions are met or at your request
```
Comment (amurKontur) [inline your_make.sample:20]: please, add an example for unexperienced with geocint users. on how it should look like

File: your_make.sample [hunk 23]
```diff
@@ -1,4 +1,32 @@
...
+
+# by default the clean target is set to serve an update of the OpenStreetMap planet dump during every run
+clean: ## [FINAL] Cleans the worktree for next nightly run. Does not clean non-repeating targets.
+	if [ -f data/planet-is-broken ]; then rm -rf data/planet-latest.osm.pbf ; fi
+	rm -rf data/planet-is-broken
+	profile_make_clean data/planet-latest-updated.osm.pbf
+
+## --------------- SAMPLE TARGET CHAIN ------------------
 
 ## This is an example makefile created to demonstrate how to create and use targets as part of the ETL process.
```
Comment (amurKontur) [inline your_make.sample:31]: add details on what is ETL

Comment (frolui) [reply your_make.sample]: fixed

File: your_make.sample (hunk 22 above)
Comment (frolui) [reply your_make.sample]: added explanation

**Review by Komzpa — APPROVED** [review]

File: scripts/set_modes.sh [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+
+# this script is using one during installation to create /public_html and /domlogs folders and set access modes
```
Comment (Lokks) [inline scripts/set_modes.sh:3]: once?

File: start_geocint.sh [hunk 25]
```diff
@@ -1,24 +1,26 @@
 #!/bin/bash
 
-# Set variables
+# it is the script, that runs the pipeline: it checks required packages, cleans and runs targets, posts info messages
```
Comment (Lokks) [inline start_geocint.sh:3]: this is the script

File: start_geocint.sh [hunk 26]
```diff
@@ -1,24 +1,26 @@
...
 . ~/config.inc.sh
-export PATH_ARRAY GENERAL_FOLDER OPTIONAL_DIRECTORIES PRIVATE_REPO_NAME PRIVATE_MAKE_NAME OSM_MAKE_NAME 
-export UPDATE_RUNNER UPDATE_OSM_LOGIC UPDATE_PRIVATE ALL_TARGETS RUN_TARGETS
-export SLACK_CHANNEL SLACK_BOT_NAME SLACK_BOT_EMOJI PGDATABASE
-export TARGET_TO_CLEAN RM_DIRECTORIES CLEAN_OPTIONALLY
+export PATH_ARRAY GENERAL_FOLDER OPTIONAL_DIRECTORIES PRIVATE_REPO_NAME
+export UPDATE_RUNNER UPDATE_OSM_LOGIC UPDATE_PRIVATE RUN_TARGETS
+export SLACK_CHANNEL SLACK_BOT_NAME SLACK_BOT_EMOJI SLACK_KEY USER_NAME
 
+# initialize the pipeline interrupt function for the case when the pipeline can't touch make.lock
```
Comment (Lokks) [inline start_geocint.sh:10]: in case when pipeline can't...

File: your_make.sample [hunk 27]
```diff
@@ -21,6 +51,7 @@ db/table/fire_stations: db/index/osm_tags_idx ## Extract centroids of fire stati
 					order by osm_id \
```
Comment (Lokks) [inline your_make.sample:51]: do we need order by here?

File: scripts/set_modes.sh (hunk 24 above)
Comment (frolui) [reply scripts/set_modes.sh:3]: yeah

**Pull request merged**